Improve DeltaList behavior for large projects

gsmet opened the pull request:
This pull request targets the 3.x branch as I think this support has been rewritten in 4.

Here is the situation prior to this patch with a massive project (~40k source files):

![Screenshot from 2025-06-03 17-17-35](https://github.com/user-attachments/assets/f8d85bc1-714d-4fba-9611-2b61c0825db0)

> 7172 samples, 3.79%

The first commit keeps the ordering as is and make sure we write the file sorted and the outputs of `DeltaList` are sorted.
The behavior is exactly the same as before.

Here is the situation after the first commit:

![Screenshot from 2025-06-03 17-17-27](https://github.com/user-attachments/assets/8fffb8be-ad5d-4c63-bb0f-9ff9a1507e54)

> 78 samples, 0.04% - basically, the only thing left is the tiny thing you can see at the far right of the previous flame graph.

The second commit is a further optimization, which might not be useful and only sorts when displaying the output. The file is written unsorted.
It doesn't change much tbh so we might as well not include the second commit.
I did the work to be comprehensive so I thought I would present it.

Here is the situation after the second commit:

![Screenshot from 2025-06-03 17-17-43](https://github.com/user-attachments/assets/cc4c74ce-0d38-4596-83b2-cad4b43e79a4)

> 59 samples, 0.04%

**Opening as a draft so that we decide what to do about the second commit: my advice would be to drop it but YMMV**

BTW, it might be possible to optimize further but given this code is readable and give good results, I think it's good enough.

Fixes https://github.com/apache/maven-compiler-plugin/issues/334

----

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

- [X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
- [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

jorsol commented on src/main/java/org/apache/maven/plugin/compiler/DeltaList.java:
```diff
@@ -33,10 +33,8 @@ final class DeltaList<E> {
-        this.removed = new ArrayList<>(oldList);
-        added.removeAll(oldList);
-        removed.removeAll(newList);
+        this.added = newList.stream().filter(i -> !oldList.contains(i)).collect(Collectors.toList());
+        this.removed = oldList.stream().filter(i -> !newList.contains(i)).collect(Collectors.toList());
```
Could it be better to use the old for loop instead of streams?

gsmet replied:
We could sure.

cstamas commented:
@desruisseaux @gnodet ping

gsmet commented:
I adjusted the commits:
- dropped the commit removing the sorts - this commit made things a bit faster but it's probably more future proof to keep the output properly sorted
- I switched from streams to for loops in `DeltaList`

desruisseaux commented:
It looks fine to me as well.

gnodet commented on src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java:
```diff
@@ -1840,19 +1840,21 @@ private boolean hasInputFileTreeChanged(IncrementalBuildHelper ibh, Set<File> in
+        Set<String> newInputFiles = inputFiles.stream()
+                .sorted()
+                .map(File::getAbsolutePath)
+                .collect(Collectors.toCollection(LinkedHashSet::new));
```
> Would it be better to use a `TreeSet` instead of sorting + `LinkedHashSet` ?

Quick lookup seems to indicate sorting + collecting the stream is slightly faster.
